### PR TITLE
feat: new simultaneous inpainter

### PR DESCRIPTION
**Pull request by steven-murray**
This puts @tyler-a-cox's new simultaneous inpainter into `hera_cal.lst_stack.averaging`.

I refactored it a little so that the algorithm itself, that acts on a single baseline, is its own function. This makes testing it much easier, since you just have to create data arrays that are shape (nnights, nfreqs).

I've put a number of test cases in (as of now, not all of them pass...)

### CODE REVIEW

**Review by jsdillon — CHANGES_REQUESTED** [review]
Just some minor comments, but this generally looks good to me. Thanks @steven-murray

File: hera_cal/lst_stack/averaging.py [hunk 1]
```diff
@@ -345,77 +347,272 @@ def reduce_lst_bins(
...
+    inpaint_bands : tuple[slice]
+        The bands to inpaint individually, as slices over frequency channels.
+    max_gap_factor : float
+        The maximum allowed gap factor.
+    max_convolved_flag_frac : float
+        The maximum allowed fraction of convolved flags.
+    use_unbiased_estimator : bool
+        Whether to use an unbiased estimator.
+    sample_cov_fraction : float
+        A factor to use to down-weight off-diagonal elements of the sample covariance.
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lst_stack/averaging.py:398]: Good point. Done.

File: hera_cal/lst_stack/averaging.py [hunk 2]
```diff
@@ -345,77 +347,272 @@ def reduce_lst_bins(
...
+    # DPSS inpainting model
+    model = np.zeros_like(stackd)
+    mask = (~stackf).astype(float)
+    avg_flgs = avg_flgs if avg_flgs is not None else np.all(stackf, axis=0)
+
+    n_nights = stackd.shape[0]
+
+    # Get median nsamples across the band
+    nsamples_by_night = np.median(stackn, axis=1, keepdims=True)
+    assert np.all(nsamples_by_night == stackn), 'This code assumes that nsamples is constant over frequency for a given night and baseline.'
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lst_stack/averaging.py:415]: Yup. Done.

File: hera_cal/lst_stack/averaging.py [hunk 3]
```diff
@@ -345,77 +347,272 @@ def reduce_lst_bins(
...
+    base_noise_var: np.ndarray,
+    df: un.Quantity['frequency'],
+    filter_centers: Sequence[float],
+    avg_flgs: np.ndarray | None = None,
+    inpaint_bands: tuple[slice] = (slice(0, None, None),),
+    max_gap_factor: float = 2.0,
+    max_convolved_flag_frac: float = 0.667,
+    use_unbiased_estimator: bool = False,
+    sample_cov_fraction: float = 0.0,
+    filter_half_widths: Sequence[float] = [0.1],
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lst_stack/averaging.py:364]: You're right -- I've removed the default.

File: hera_cal/lst_stack/averaging.py [hunk 4]
```diff
@@ -345,77 +347,272 @@ def reduce_lst_bins(
...
-    .. math:: m = \text{FourierFilter}(\bar{d})
-
-    Finally, it will do a weighted average of the data and the model:
+def average_and_inpaint_simultaneously_single_bl(
+    freqs: np.ndarray,
+    stackd: np.ndarray,
+    stackf: np.ndarray,
+    stackn: np.ndarray,
+    base_noise_var: np.ndarray,
+    df: un.Quantity['frequency'],
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lst_stack/averaging.py:356]: Made it flexible (if a float, assume it's in Hz) and added docs.

File: hera_cal/lst_stack/averaging.py [hunk 5]
```diff
@@ -345,77 +347,272 @@ def reduce_lst_bins(
...
+    inpainted_mean = np.zeros(len(freqs), dtype=stackd.dtype)
+    total_nsamples = np.zeros(len(freqs), dtype=float)
+
+    for band in inpaint_bands:
+        # if the band is already entirely flagged for all nights, continue
+        if np.all(stackf[:, band]):
+            continue
+
+        # if there are too-large flag gaps even after a simple LST-stacking, continue
+        max_allowed_gap_size = max_gap_factor * filter_half_widths[0]**-1 / df.value
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lst_stack/averaging.py:427]: Done.

File: hera_cal/lst_stack/averaging.py [hunk 6]
```diff
@@ -345,77 +347,272 @@ def reduce_lst_bins(
...
+    max_gap_factor : float
+        The maximum allowed gap factor.
+    max_convolved_flag_frac : float
+        The maximum allowed fraction of convolved flags.
+    use_unbiased_estimator : bool
+        Whether to use an unbiased estimator.
+    sample_cov_fraction : float
+        A factor to use to down-weight off-diagonal elements of the sample covariance.
+    filter_half_widths : Sequence[float]
+        The half-widths of the DPSS filters.
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lst_stack/averaging.py:400]: Added that info to the docstring.

File: hera_cal/lst_stack/averaging.py [hunk 7]
```diff
@@ -345,77 +347,272 @@ def reduce_lst_bins(
...
+        The stacked flags, shape (n_nights, nfreq).
+    stackn : np.ndarray
+        The stacked nsamples, shape (n_nights, nfreq).
+    base_noise_var : np.ndarray
+        The expected noise variance for each night, shape (n_nights, nfreq).
+    df : un.Quantity['frequency']
+        The frequency resolution.
+    filter_centers : Sequence[float]
+        The centers of the DPSS filters.
+    avg_flgs : np.ndarray | None
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_cal/lst_stack/averaging.py:387]: You're right. I've updated it.

**Review by jsdillon — APPROVED** [review]

**Pull request merged**